core/domain/exp_domain.py

```diff
@@ -2213,6 +2246,167 @@ def _convert_states_v28_dict_to_v29_dict(cls, states_dict):
             state_dict['solicit_answer_details'] = False
         return states_dict
 
+    @classmethod
+    def _convert_states_v29_dict_to_v30_dict(cls, states_dict):
+        """Converts from version 29 to 30. Version 30 added image assets field
+        in state model. This funtion is extracting HTMl of state and do 3
+        operations on each HTML content. The 3 operations and their sequence
+        were.
+        1) Add image_id in each image tag present in the HTML.
+        2) Extract image info of each image from HTML.
+        3) Remove filepath of each image present in the HTML.
+
+         Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        image_counter = 0
+        for state_dict in states_dict.itervalues():
+            image_assets = {}
+            image_mapping = {}
+            image_info_list = []
+
+            # Adding image_id, getting image info and removing filepath from
+            # image tag present in state content.
+            content_html = state_dict['content']['html']
+            (new_content_html, image_counter) = (
+                html_validation_service.add_image_id_in_image_tag(
```

Comment:
Let's not create an HTML content which is neither follows new template nor the old.
- get image info if the image exists. (Just extract the data do not remove it.)
- Replace image file_path with image_id (pass filepath and corresponding image_id)
- Populate the image_info and image_id in the image_assets field.

Will this be fine?

Reply:
One point to say, why you were not deleting ```filepath```, why to replace?

And can you tell me what is the ```BeautifulSoup``` API for replacing attribute?

Reply:
> Replace image file_path with image_id (pass filepath and corresponding image_id)

The name of the method will contain `replace` but internally it will check the file_pat and deletes the file path then adds the corresponding image_id.

> And can you tell me what is the BeautifulSoup API for replacing attribute?

I'm not sure, you can check this in one web-search. (i.e, whether such functionality exists or not)

Reply:
Thanks, @DubeySandeep, I searched it,  but I think it's better to delete ```filepath``` and add ```image-id```.
And I am fine with this

> The name of the method will contain `replace` but internally it will check the file_pat and deletes the file path then adds the corresponding image_id.